### ISSUE TITLE
Please use license consistently in the release jar files

### ISSUE THREAD

**Issue by vlsi** [ben-manes/caffeine#325]
1) Please add `Bundle-License: Apache-2.0` manifest entry (where `Apache-2.0` is SPDX identifier, see https://osgi.org/specification/osgi.core/7.0.0/framework.module.html#framework.module-bundle-license )
Well, current `pom.xml` tags are quite obvious (even though you might want to replace `<licence><name>` with SPDX id of `Apache-2.0`), however adding explicit `Bundle-License` would simplify automatic analysis.

2) Please add LICENSE and NOTICE files to the produced jar files (and probably to the source repository).
The thing is the license requires re-distributions to reproduce copyright notice, however it is really hard to do that in an automatic way, taking into account `caffeine.jar` is often taken from Maven Central. The jar file just misses NOTICE file, so it is really hard to reproduce copyright notice and track it for all caffeine upgrades.

The suggested file locations are `META-INF/LICENSE` and `META-INF/NOTICE`.

The relevant link is http://www.apache.org/legal/src-headers.html#notice

PS I'm implementing an automatic license analysis for Apache JMeter, and I noticed NOTICE is not distributed by caffeine.

**Comment by ben-manes** [comment]
Sounds good, just requires some Gradle fiddling. PR's welcome or I'll try to do it whenever.

`NOTICE` is optional, so you shouldn't be failing if that is missing. In legalize, `should` is a recommendation, not a requirement.

The `pom.xml` is usually the preferred check prior to scanning for other files, so I think you should be okay as is?

**Comment by ben-manes** [comment]
btw, `<license><name>` is correctly the full string for Maven's pom. The SPIX id would be an incorrect replacement. I'm fine adding the OSGi manifest entry, etc. though.

**Comment by vlsi** [comment]
>The pom.xml is usually the preferred check prior to scanning for other files, so I think you should be okay as is?

Frankly speaking, I would consider `pom.xml` to be the last resort, since `Bundle-License` is somehow settled on SPDX.

> NOTICE is optional, so you shouldn't be failing if that is missing

> btw, <license><name> is correctly the full string for Maven's pom. The SPIX id would be an incorrect replacement

Why do you think so?
Of course `<license><name>` might be custom for `non-trivial` licenses, however for simple cases like `Apache-2.0` I think it makes more sense to just use SPDX id in `<name>` to avoid ever-going debates on the exact spelling  (e.g. "v2.0" vs "2.0" and so on).

There's a ticket to add machine-readable license id to pom.xml, however, as you might guess it won't be resolved in the nearest future: https://issues.apache.org/jira/browse/MNG-6677

**Comment by vlsi** [comment]
> NOTICE is optional, so you shouldn't be failing if that is missing. In legalize, should is a recommendation, not a requirement.

I'm not a lawyer, but AL reads `must` in a `redistribution` section: https://github.com/ben-manes/caffeine/blob/master/LICENSE#L101-L103

I guess it is just just a way simpler to add a relevant NOTICE file rather than debate which cases require NOTICE and which do not.

As a bonus, your work will be acknowledged in redistributions which is nice in my point of view.

**Comment by vlsi** [comment]
PS. I've no strong opinion on `<license><name>`, and the most severe issue for me is the absence of LICENSE/NOTICE in the jar file, as my tool could somehow infer AL2 from your current `pom.xml`.

**Comment by ben-manes** [comment]
http://www.apache.org/legal/src-headers.html#notice
>Every Apache distribution should include a NOTICE file in the top directory, along with the standard LICENSE file.

The requirements of the consumer are dictated by the `LICENSE` file. However, the convention for an Apache project structure is determined by their policies, where `NOTICE` is optional.

https://issues.apache.org/jira/browse/MPIR-382
> There is only one spelling: Apache License, Version 2.0. The proper solution is to report this upstream.

Adding missing items is good, so happy to do that. I am only hesitant when modify, where I think we are doing it correctly atm.

**Comment by vlsi** [comment]
> vlsi: but AL reads must in a redistribution section: https://github.com/ben-manes/caffeine/blob/master/LICENSE#L101-L103

Ah, you are right.
Sorry for the noise. That section means "if you use AL2.0 source, don't remove copyright headers from the source form".

On the other hand you would get a bit of appreciation if you redistribute NOTICE (you would probably need it update from year to year though :) )

**Comment by ben-manes** [comment]
Yep, happy to add `NOTICE`, `Bundle-License`, etc. Thanks. :)

**Comment by vlsi** [comment]
@ben-manes , 

```
Bundle-License ::= '<<EXTERNAL>>' | 
                      ( license ( ',' license ) * )
license        ::= license-identifier ( ';' license-attr ) *
license-attr   ::= description | link
description    ::= 'description' '=' string
link           ::= 'link' '=' <url>
```

I'm not sure `'Bundle-License': 'http://www.apache.org/licenses/LICENSE-2.0'` qualifies.

AFAIK here a the valid combos:

```
Bundle-License: Apache-2.0
Bundle-License: Apache-2.0;link=http://www.apache.org/licenses/LICENSE-2.0
```

**Comment by ben-manes** [comment]
hmm, I looked at what Maven produces for major dependencies and it was consistently what they do. Here are some from dependencies used in via tests which use this pattern. I don't see any that follow your examples,

- guice-4.2.2
- guava-28.0
- testng-6.14.3
- commons-lang3-3.9
- byte-buddy-1.9.10
- objenesis-2.6
- jackson-core-2.8.11

**Comment by ben-manes** [comment]
Even Apache Felix, and OSGi container, uses the http address. I assume it must be from the OSGi plugin from Maven. It seems less risky to parse if we follow what everyone else has, regardless of whether the grammar states.

**Comment by vlsi** [comment]
Oh, there's 

>...Alternatively, the identifier can be the canonical URL of the license, i

**Comment by ben-manes** [comment]
great! I changed it slightly to be `https` just to drop `http://` from the build files, as it is provided over TLS.

**Comment by ben-manes** [comment]
Released in v2.8
